core/client/src/block_builder/block_builder.rs

```diff
@@ -115,6 +117,15 @@ where
 		})
 	}
 
+	/// Set the digest of `self` to `digest`.
+	pub fn push_digest(&mut self, digest: DigestFor<Block>) {
+		assert!(self.header.digest().logs().len() == 0, "We never call this method if there are existing digests; qed");
```

Comment:
our codebase style avoids these kinds of assertions, favoring to return some kind of error code instead. "prove or remove" can't reasonably be extended to assumptions outside the module -- this code could be called from anywhere.

As mentioned above, I think it would be safer to do the `set_inherent_digest` stuff when we _create_ the block builder, which naturally preserves this invariant.

Reply:
I agree.